sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsTestUtils.scala

```diff
@@ -214,14 +214,16 @@ trait SQLMetricsTestUtils extends SQLTestUtils {
   protected def testSparkPlanMetrics(
       df: DataFrame,
       expectedNumOfJobs: Int,
-      expectedMetrics: Map[Long, (String, Map[String, Any])]): Unit = {
+      expectedMetrics: Map[Long, (String, Map[String, Any])],
+      enableWholeStage: Boolean = false): Unit = {
     val expectedMetricsPredicates = expectedMetrics.mapValues { case (nodeName, nodeMetrics) =>
       (nodeName, nodeMetrics.mapValues(expectedMetricValue =>
         (actualMetricValue: Any) => {
           actualMetricValue.toString.matches(expectedMetricValue.toString)
         }))
     }
-    testSparkPlanMetricsWithPredicates(df, expectedNumOfJobs, expectedMetricsPredicates)
+    testSparkPlanMetricsWithPredicates(df, expectedNumOfJobs, expectedMetricsPredicates,
```

Comment:
Is it not enough just to run tests here with both modes: codegen and non-codegen?
```
Seq(true, false).foreach { codegenEnabled =>
  testSparkPlanMetricsWithPredicates(df, expectedNumOfJobs, expectedMetricsPredicates, codegenEnabled)
}
```

Reply:
Ah, the same comment there.... https://github.com/apache/spark/pull/28173#discussion_r406877479
Could you check this, @sririshindra

Reply:
@maropu 
The expectedMetricsPredicates can vary with codegenEnabled=true and codegenEnabled=false. 

Also the test "WholeStageCodegen metrics"  needs to run testSparkPlanMetricsWithPredicates method with only codegenEnabled=true. So if we change the method to run with both configurations then the test "WholeStageCodegen metrics"  will fail. 

The dataframe (variable df) needs to fresh for every call of testSparkPlanMetricsWithPredicates. The dataframe should not be reused or else the metrics will change causing the tests to fail.

Reply:
But, we should avoid duplicate code where possible.
How about this?
```
  protected def testSparkPlanMetrics(
      df: _ => DataFrame,  // To make `df` fresh as you said.
      expectedNumOfJobs: Int,
      expectedMetrics: Map[Long, (String, Map[String, Any])],
      testWithWholeStage: Boolean = false): Unit = {
    val expectedMetricsPredicates = expectedMetrics.mapValues { case (nodeName, nodeMetrics) =>
      (nodeName, nodeMetrics.mapValues(expectedMetricValue =>
        (actualMetricValue: Any) => {
          actualMetricValue.toString.matches(expectedMetricValue.toString)
        }))
    }
    testSparkPlanMetricsWithPredicates(
      df(), expectedNumOfJobs, expectedMetricsPredicates, false)

    // If testWithWholeStage=true, run the second test with codegen enabled
    if (testWithWholeStage) {
      testSparkPlanMetricsWithPredicates(
        df(), expectedNumOfJobs, expectedMetricsPredicates, true)
    }
  }
```

Reply:
Interesting idea to make the 'df' fresh.
However we still have the problem of expectedMetricsPredicates argument being different for most of the tests when codegen=true and codegen=false. Since the arguments to the testSparkPlanMetricsWithPredicates method could be different when codegen=true and codegen=false, we probably couldn’t avoid calling the testSparkPlanMetrics method twice with the corresponding arguments.

I agree with you that we should minimize the code duplication. I just pushed another commit to further refactor the code to minimize code duplication. Could you please take a look and let me know what do you think.